src/metaopt/algo/space.py

```diff
@@ -0,0 +1,536 @@
+
+        """
+        prior_low, prior_high = super(Real, self).interval(alpha)
+        return (max(prior_low, self._low), min(prior_high, self._high))
+
+    def sample(self, n_samples=1, seed=None):
+        """Draw random samples from `prior`.
+
+        .. seealso:: `Dimension.sample`
+
+        """
+        samples = []
+        for _ in range(n_samples):
+            for _ in range(4):
+                sample = super(Real, self).sample(1, seed)
+                if sample[0] not in self:
+                    nice = False
+                    continue
+                nice = True
+                samples.extend(sample)
+                break
+            if not nice:
+                raise ValueError("Improbable bounds: (low={0}, high={1}). "
+                                 "Please make interval larger.".format(self._low, self._high))
+
+        return samples
+
+
+class Integer(Real, _Discrete):
+    """Subclass of `Dimension` for representing integer parameters.
+
+    Attributes
+    ----------
+    name : str
+    type : str
+    prior : `scipy.stats.distributions.rv_generic`
+    shape : tuple
+       See Attributes of `Dimension`.
+
+    """
+
+    def __contains__(self, point):
+        """Check if constraints hold for this `point` of `Dimension`.
+
+        :param x: a parameter corresponding to this `Dimension`.
+        :type x: numeric or array-like
+
+        `Integer` will check whether `point` contains only integers.
+
+        """
+        point_ = numpy.asarray(point)
+        if not numpy.all(numpy.equal(numpy.mod(point_, 1), 0)):
+            return False
+        return super(Integer, self).__contains__(point)
+
+
+class Categorical(_Discrete):
+    """Subclass of `Dimension` for representing integer parameters.
+
+    Attributes
+    ----------
+    name : str
+    type : str
+    prior : `scipy.stats.distributions.rv_generic`
+    shape : tuple
+       See Attributes of `Dimension`.
+    categories : tuple
+       A set of unordered stuff to pick out from, except if enum
+
+    """
+
+    def __init__(self, name, categories, **kwargs):
+        """Search space dimension that can take on categorical values.
+
+        Parameters
+        ----------
+        name : str
+           See Parameters of `Dimension.__init__`.
+        categories : dict or other iterable
+           A dictionary would associate categories to probabilities, else
+           it assumes to be drawn uniformly from the iterable.
+        kwargs : dict
+           See Parameters of `Dimension.__init__` for general.
+
+        """
+        if isinstance(categories, dict):
+            self.categories = tuple(categories.keys())
+            self._probs = tuple(categories.values())
+        else:
+            self.categories = tuple(categories)
+            self._probs = tuple(numpy.tile(1. / len(categories), len(categories)))
+
+        prior = distributions.rv_discrete(values=(list(range(len(self.categories))),
```

Comment:
I stand by my point here. I don't see the advantage of going through rv_discrete. Using `numpy.random.choices` would be simpler, without a need for any `_inverse` and `_check`.